File: src/test/java/dev/bothin/testcontainers/java/ContainerConfiguration.java
package dev.bothin.testcontainers.java;

import dev.bothin.testcontainers.TestContainerConfiguration;
import org.testcontainers.containers.wait.strategy.Wait;

public class ContainerConfiguration implements TestContainerConfiguration {

    private MqttContainer mqttContainer;

    public ContainerConfiguration() {
        mqttContainer = new MqttContainer("eclipse-mosquitto");
        mqttContainer.withExposedPorts(1883).waitingFor(Wait.forLogMessage(".*Config loaded from.*", 1));
    }

    @Override
    public void startAll() {
        mqttContainer.start();
    }

    @Override
    public void afterStart() {
        System.setProperty("MQTT_HOST", mqttContainer.getContainerIpAddress());
        System.setProperty("MQTT_PORT", mqttContainer.getMappedPort(1883).toString());
    }

    @Override
    public void stopAll() {
        mqttContainer.stop();
    }
}

File: src/test/kotlin/dev/bothin/testcontainers/ContainerConfiguration.kt
package dev.bothin.testcontainers

import org.testcontainers.containers.GenericContainer
import org.testcontainers.containers.wait.strategy.Wait


class ContainerConfiguration : TestContainerConfiguration {

    private val mqttContainer = MqttContainer("eclipse-mosquitto")
        .withExposedPorts(1883)
        .waitingFor(Wait.forLogMessage(".*Config loaded from.*", 1))

    override fun startAll() {
        mqttContainer.start()
    }

    override fun afterStart() {
        System.setProperty("MQTT_HOST", mqttContainer.containerIpAddress)
        System.setProperty("MQTT_PORT", mqttContainer.getMappedPort(1883).toString())
    }

    override fun stopAll() {
        mqttContainer.stop()
    }

}

open class KGenericContainer(imageName: String) : GenericContainer<KGenericContainer>(imageName)

class MqttContainer(imageName: String) : KGenericContainer(imageName)